File: test-agent-v2/tool_list_files.py
"""The list_files tool."""

import json
import os
from typing import Any

from tool import ToolDefinition


def list_files(args: dict[str, Any]) -> str:
    dir_path = args.get("path") or "."

    files: list[str] = []
    for root, dirnames, filenames in os.walk(dir_path):
        for name in dirnames:
            rel = os.path.relpath(os.path.join(root, name), dir_path)
            files.append(rel + "/")
        for name in filenames:
            rel = os.path.relpath(os.path.join(root, name), dir_path)
            files.append(rel)

    return json.dumps(files)


ListFilesTool = ToolDefinition(
    name="list_files",
    description=(
        "List files and directories at a given path. If no path is provided, "
        "lists files in the current directory."
    ),
    input_schema={
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": (
                    "Optional relative path to list files from. Defaults to "
                    "current directory if not provided."
                ),
            },
        },
    },
    function=list_files,
)

File: test-agent/tool_list_files.go
// This file contains functions for listing files.
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
)

var ListFilesTool = ToolDefinition{
	Name:        "list_files",
	Description: "List files and directories at a given path. If no path is provided, lists files in the current directory.",
	InputSchema: map[string]any{
		"type": "object",
		"properties": map[string]any{
			"path": map[string]any{
				"type":        "string",
				"description": "Optional relative path to list files from. Defaults to current directory if not provided.",
			},
		},
	},
	Function: listFiles,
}

type listFilesInput struct {
	Path string `json:"path,omitempty"`
}

func listFiles(input json.RawMessage) (string, error) {
	in := listFilesInput{}
	if len(input) > 0 {
		if err := json.Unmarshal(input, &in); err != nil {
			return "", err
		}
	}

	dir := "."
	if in.Path != "" {
		dir = in.Path
	}

	var files []string
	err := filepath.Walk(dir, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(dir, p)
		if err != nil {
			return err
		}
		if rel != "." {
			if info.IsDir() {
				files = append(files, rel+"/")
			} else {
				files = append(files, rel)
			}
		}
		return nil
	})
	if err != nil {
		return "", err
	}

	out, err := json.Marshal(files)
	if err != nil {
		return "", err
	}
	return string(out), nil
}
